frontend/gateway/grpcclient/client.go

```diff
@@ -71,16 +73,22 @@ func convertRef(ref client.Reference) (string, error) {
 	return r.id, nil
 }
 
-func Run(ctx context.Context, f client.BuildFunc) (retError error) {
+func RunFromEnvironment(ctx context.Context, f client.BuildFunc) error {
 	client, err := current()
 	if err != nil {
 		return errors.Wrapf(err, "failed to initialize client from environment")
 	}
+	return client.Run(ctx, f)
+}
 
-	res, err := f(ctx, client)
-
-	export := client.caps.Supports(pb.CapReturnResult) == nil
+func (c *grpcClient) Run(ctx context.Context, f client.BuildFunc) (retError error) {
+	// TBD: in a client-side gateway situation if !export and f() returns an error, how to kick the parent `Solve` into exiting?
```

Comment:
I'm a bit unclear on this, although I've moved the call to `f` later such that the `if export defer` stuff runs even if it fails I'm not convinced that is necessary -- the return from `f` would cause the surrounding `errgroup` to cancel which _ought_ to cancel the other goroutine doing the `Solve` -- but I failed to trace the code far enough to confirm that cancelling the context would really cancel the gRPC.

Also the lack of similar handling in the `!export` case concerns me. Not sure what I should do there (depending onthe above, perhaps the answer is "nothing").

Reply:
> I think we can squash retError to nil in the defer, (unless c.client.Return itself fails, we should propagate that error still).

No, we should still return it here. For example, a frontend should use it to return an error code as well. We can skip it in the client if we have the main error instead

Reply:
We have `cb` in `Client.Build` capture any `err` but still `return nil`? Then we take the result of `c.solve` and if it is an err use that, but if it is not an error use the `err` capturerd from the `cb` instead? Something like:
```go
	var cbErr
	cb := func(ref string, s *session.Session) error {
		g, cbErr := grpcclient.New(ctx, feOpts, s.ID(), product, c.gatewayClientForBuild(ref), gworkers)
		if cbErr != nil {
			return nil
		}

		cbErr = g.Run(ctx, buildFunc)
		return nil
	}

	if err := c.solve(ctx, nil, cb, opt, statusChan); err != nil {
		return err
	}
	return cbErr
```
???

Reply:
This wasn't quite right, I've pushed a commit along these lines though.

Reply:
To make this slightly safer we could use a timeout. So if cb errors, and 5 seconds passes it will cancel the main context. This code can also be in to `solve` itself, not handled by the closure.

Reply:
Done. I left it in a separate commit for now.